verification/curator-service/src/controllers/sources.ts

```diff
@@ -7,37 +7,71 @@ import { Source } from '../model/source';
  */
 export const list = async (req: Request, res: Response): Promise<void> => {
     const docs = await Source.find({});
-    res.send(docs);
+    res.json(docs);
 };
 
 /**
  * Get a single source.
  */
-export const get = (req: Request, res: Response): void => {
-    res.sendStatus(501);
-    return;
+export const get = async (req: Request, res: Response): Promise<void> => {
+    const doc = await Source.findById(req.params.id);
+    if (!doc) {
+        res.sendStatus(404);
+        return;
+    }
+    res.json(doc);
 };
 
 /**
  * Update a single source.
  */
-export const update = (req: Request, res: Response): void => {
-    res.sendStatus(501);
-    return;
+export const update = async (req: Request, res: Response): Promise<void> => {
+    try {
+        const source = await Source.findByIdAndUpdate(req.params.id, req.body, {
+            // Return the udpated object.
+            new: true,
+            runValidators: true,
+        });
+        if (!source) {
+            res.sendStatus(404);
+            return;
+        }
+        res.json(source);
+    } catch (err) {
+        if (err.name === 'ValidationError') {
+            res.status(422).json(err.message);
+            return;
+        }
+        res.status(500).json(err.message);
+        return;
+    }
 };
 
 /**
  * Create a single source.
  */
-export const create = (req: Request, res: Response): void => {
-    res.sendStatus(501);
-    return;
+export const create = async (req: Request, res: Response): Promise<void> => {
+    try {
+        const source = await Source.create(req.body);
+        res.location(`/sources/${source.id}`);
```

Comment:
We should be returning the created object; is that done automagically via middleware by virtue of setting this?

I'd assume the opposite, that location were the potential middleware automagic.

Reply:
Hmm that's kind of an unofficial REST convention, returning the location of the created object instead of the whole object.
I don't mind returning a 200 and the created object as json if you prefer, the location header containing the url is how I've done that in the past that's all.
**I think the point is that clients shouldn't have to build url paths, they should just navigate with urls given by the server (here follow the location header value).**

- https://restfulapi.net/http-methods/#post
- stackoverflow is more nuanced, there seem to be 3 school of thoughts (use location, return full object, do both)